rust/kernel/src/sync.rs

```diff
@@ -0,0 +1,145 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use core::cell::UnsafeCell;
+use core::fmt;
+use core::ops::{Deref, DerefMut};
+
+use crate::bindings;
+
+pub struct Mutex<T: ?Sized> {
+    lock: bindings::mutex,
+    data: UnsafeCell<T>,
+}
+
+impl<T> Mutex<T> {
+    /// Create a new Mutex
+    pub fn new(data: T, _name: &'static str) -> Self {
+        let lock = bindings::mutex::default();
+        // TODO: write mutex debug traces like .magic.
+        // TODO: use name in the debug version
+
+        Self {
+            data: UnsafeCell::new(data),
+            lock,
+        }
+    }
+}
+
+impl<T: ?Sized> Mutex<T> {
+    /// acquire a lock on the mutex
+    /// # unsafe
+    /// This is unsafe, as it returns a second lock if one is already help by the current process
+    // with CONFIG_DEBUG_LOCK_ALLOW mutex_lock is a macro, which calls mutex_lock_nested(&mutex, 0)
+    #[cfg(CONFIG_DEBUG_LOCK_ALLOC)]
+    pub unsafe fn lock<'a>(&'a self) -> MutexGuard<'a, T> {
+        unsafe {
+            bindings::mutex_lock_nested(
```

Comment:
You need to call `mutex_init` before you can safely use it. Where is it being called?

Reply:
For some mutexes (I think it was in the rtlink code) I don't saw it called anywhere. So I thought it works without calling mutex_unit if you set the head_list

Reply:
> For some mutexes (I think it was in the rtlink code) I don't saw it called anywhere.

In such cases they were probably defined with `DEFINE_MUTEX`, which allows them to skip `mutex_init`.

> So I thought it works without calling mutex_unit if you set the head_list

Where are you setting the head of the list?

In any case, check out [Generic Mutex Subsystem](https://www.kernel.org/doc/html/latest/locking/mutex-design.html), in particular:
_A mutex must only be initialized via the API (see below)._